shell: add obscured input mode

plskeggs opened the pull request:
Add flag to shell_flags to control echo as obscure character (for passwords) or not. Add setter and getter
functions. Modify shell echo to use this flag if set.

Add public setter functions for color mode and obscure mode.

Update sample to show how to use this for login and logout commands.

Signed-off-by: Pete Skeggs <peter.skeggs@nordicsemi.no>

plskeggs committed:
shell: add obscured input mode

Add flag to shell_flags to control echo as obscure
character (for passwords) or not. Add setter and getter
functions. Modify shell echo to use this flag if set.

Also add public setter for color mode and obscure mode.

Signed-off-by: Pete Skeggs <peter.skeggs@nordicsemi.no>

plskeggs committed:
shell: sample: add login example

Using the new obscured user shell input feature,
provide examples of using it in login and logout commands.

Signed-off-by: Pete Skeggs <peter.skeggs@nordicsemi.no>

nordic-krch commented on samples/subsys/shell/shell_module/src/main.c:
```diff
@@ -204,6 +257,10 @@ SHELL_CMD_ARG_REGISTER(version, NULL, "Show kernel version", cmd_version, 1, 0);
 
 void main(void)
 {
+#if defined(CONFIG_SHELL_START_OBSCURED)
```
you could use `if (IS_ENABLED(CONFIG_SHELL_START_OBSCURED))` and avoid multiple ifdefs. Using `SHELL_COND_CMD_ARG_REGISTER`.

plskeggs replied:
done

nordic-krch reviewed (commented):
Looks good. Can you also update `sample.yaml` to ensure that configuration is compiled?

jakub-uC commented:
@plskeggs Thank you for this feature. Please also mention this feature in the shell documentation:
https://github.com/zephyrproject-rtos/zephyr/blob/master/doc/reference/shell/index.rst

plskeggs committed:
doc: shell update with obscured input

Update the shell documentation with the obscured
input feature.

Signed-off-by: Pete Skeggs <peter.skeggs@nordicsemi.no>

plskeggs commented:
added docs

jharris-intel commented on include/shell/shell.h:
```diff
@@ -1034,6 +1035,30 @@ int shell_execute_cmd(const struct shell *shell, const char *cmd);
+ * @retval 0 if set.
+ * @retval -EINVAL if shell is NULL.
+ */
+int shell_obscure_set(const struct shell *shell, bool obscure);
```
I'd somewhat prefer if these APIs handled nesting properly somehow.

As-is, if you have a function that tries to temporarily have obscured input or disabled colors that's called from a context that already has obscured input or disabled colors you're going to have a bad time.

plskeggs replied:
Well, how about either also adding getters, so the application can restore the previous setting?

jharris-intel replied:
> Well, how about also adding getters, so the application can restore the previous setting?

It'd ideally be a setter+return previous, like how `arch_irq_lock` does it.

Otherwise there's a race condition if another thread preempts you between you getting the old value and setting a new one.

plskeggs replied:
I've added a new commit with changes to enable atomic set of flags and return the previous value.  I went ahead and added a few more public setters using this.

plskeggs committed:
shell: make flag modification atomic

Add Z_SHELL_SET_FLAG_ATOMIC macro to use technique from
shell_process() to atomically set and return the previous
value.

Change all void z_flag_*_set() functions to bool z_flag_*_set()
and use macro to return previous value after setting.

Modify shell_process() to use this modified z_flag_processing_set()
function.

Reorder flags to keep last_nl on byte boundary.

Add public setters for insert_mode, echo, and mode_delete for
completeness.

Signed-off-by: Pete Skeggs <peter.skeggs@nordicsemi.no>

plskeggs commented:
Anyone have any other comments on this PR?
